src/shogun/classifier/Perceptron.cpp

```diff
@@ -78,44 +53,39 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	auto pb = SG_PROGRESS(range(max_iter));
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+	m_output = SGVector<float64_t>(features->get_num_vectors());
+}
+
+void CPerceptron::iteration()
+{
+	bool converged = true;
+	SGVector<float64_t> w = get_w();
+
+	auto labels = binary_labels(m_labels)->get_int_labels();
+	auto iter_train_labels = labels.begin();
+	auto iter_output = m_output.begin();
+
+	for (const auto& v : DotIterator(features))
 	{
-		COMPUTATION_CONTROLLERS
-		converged=true;
-		auto iter_train_labels = train_labels.begin();
-		auto iter_output = output.begin();
-		for (const auto& v : DotIterator(features))
+		const auto true_label = *(iter_train_labels++);
+		auto& predicted_label = *(iter_output++);
+
+		predicted_label = v.dot(w) + bias;
```

Comment:
I have a final question here:
`predicted_label` is a scalar, and it seems that the previous iterations ouputs are not used in the current iteration. Why do we need to store the vector of them?
Couldnt we just not have the `m_output` and instead just define a local variable in the iteration method?